cpp/src/arrow/compute/exec/asof_join_benchmark.cc

```diff
@@ -0,0 +1,1023 @@
+      new RecordBatchFileReaderWrapper(in_reader));
+
+  auto schema = reader->schema();
+  // we assume there is a time field represented in uint64, a key field of int32, and the
+  // remaining fields are float64.
+  size_t row_size =
+      sizeof(_Float64) * (schema->num_fields() - 2) + sizeof(int64_t) + sizeof(int32_t);
+  auto batch_gen = *arrow::compute::MakeReaderGenerator(
+      std::move(reader), arrow::internal::GetCpuThreadPool());
+  int64_t rows = in_reader->CountRows().ValueOrDie();
+  // cout << "create source("<<filename<<")\n";
+  return {*arrow::compute::MakeExecNode(
+              "source",    // registered type
+              plan.get(),  // execution plan
+              {},          // inputs
+              arrow::compute::SourceNodeOptions(
+                  std::make_shared<arrow::Schema>(*schema),  // options, )
+                  batch_gen)),
+          rows, in_reader->num_record_batches(), row_size * rows};
+}
+
+static void TableJoinOverhead(benchmark::State& state, std::string left_table,
+                              std::vector<std::string> right_tables,
+                              std::string factory_name, ExecNodeOptions& options) {
+  const std::string data_directory = "./benchmark_data/";
+  DoSetup();
+  ExecContext ctx(default_memory_pool(), arrow::internal::GetCpuThreadPool());
+  // std::cout << "beginning test for " << left_table << " and " << right_tables[0] << " "
+  // << factory_name << std::endl; std::cout << "starting with " <<
+  // ctx.memory_pool()->bytes_allocated() << std::endl;
+  int64_t rows;
+  int64_t bytes;
+  for (auto _ : state) {
+    state.PauseTiming();
+
+    ASSERT_OK_AND_ASSIGN(std::shared_ptr<arrow::compute::ExecPlan> plan,
+                         ExecPlan::Make(&ctx));
+    std::shared_ptr<arrow::fs::FileSystem> fs =
+        std::make_shared<arrow::fs::LocalFileSystem>();
+    arrow::compute::ExecNode* left_table_source;
+    int64_t left_table_rows;
+    int left_table_batches;
+    size_t left_table_bytes;
+    tie(left_table_source, left_table_rows, left_table_batches, left_table_bytes) =
+        make_arrow_ipc_reader_node(plan, fs, data_directory + left_table);
+    std::vector<ExecNode*> inputs = {left_table_source};
+    int right_hand_rows = 0;
+    int64_t right_hand_bytes = 0;
+    for (std::string right_table : right_tables) {
+      arrow::compute::ExecNode* right_table_source;
+      int64_t right_table_rows;
+      int right_table_batches;
+      size_t right_table_bytes;
+      tie(right_table_source, right_table_rows, right_table_batches, right_table_bytes) =
+          make_arrow_ipc_reader_node(plan, fs, data_directory + right_table);
+      inputs.push_back(right_table_source);
+      right_hand_rows += right_table_rows;
+      right_hand_bytes += right_table_bytes;
+    }
+    rows = left_table_rows + right_hand_rows;
+    bytes = left_table_bytes + right_hand_bytes;
+    ASSERT_OK_AND_ASSIGN(arrow::compute::ExecNode * asof_join_node,
```

Comment:
Minor nit: `join_node` instead of `asof_join_node` since this function is used in both paths?